sqlalchemy_exasol/merge.py

```diff
@@ -11,115 +11,136 @@
         if where is not None:
-            self._insert_where = self._append_where(self._insert_where, where)
+            self._insert_where = Merge._append_where(self._insert_where, where)
 
     @_generative
     def delete(self, where=None):
         self._merge_delete = True
         if self._merge_update_values is None and where is not None:
-            self._delete_where = self._append_where(self._delete_where, where)
+            self._delete_where = Merge._append_where(self._delete_where, where)
 
-    def _append_where(self, where, addition):
-        if where is not None:
-            where = and_(where, addition)
+    @staticmethod
+    def _append_where(where, addition):
+        if where is None:
+            return addition
+        return and_(where, addition)
+
+    def _is_merge_update(self):
+        return self._merge_update_values is not None
+
+    def _is_merge_delete(self):
+        return not self._is_merge_update() and self._merge_delete
+
+    def _is_merge_insert(self):
+        return self._merge_insert_values is not None
+
+    def _compile_update(self, compiler):
+        columns = crud._get_crud_params(compiler, self._merge_update_values)
+        sql = ""
+        sql += "\nWHEN MATCHED THEN UPDATE SET "
+        sql += ", ".join(compiler.visit_column(c[0]) + "=" + c[1] for c in columns)
+        if self._merge_delete:
+            sql += "\nDELETE "
+            if self._delete_where is not None:
+                sql += " WHERE %s" % compiler.process(self._delete_where)
         else:
-            where = addition
-        return where
+            if self._update_where is not None:
+                sql += " WHERE %s" % compiler.process(self._update_where)
+        return sql
+
+    def _compile_delete(self, compiler):
+        sql = "\nWHEN MATCHED THEN DELETE "
+        if self._delete_where is not None:
+            sql += "WHERE %s" % compiler.process(self._delete_where)
+        return sql
+
+    def _compile_insert(self, compiler):
+        columns = crud._get_crud_params(compiler, self._merge_insert_values)
+        sql = "\nWHEN NOT MATCHED THEN INSERT "
+        sql += "(%s) " % ", ".join(compiler.visit_column(c[0]) for c in columns)
+        sql += "VALUES (%s) " % ", ".join(c[1] for c in columns)
+        if self._insert_where is not None:
+            sql += "WHERE %s" % compiler.process(self._insert_where)
+        return sql
+
+    def visit(self, compiler):
+        sql = "MERGE INTO %s " % compiler.process(self._target_table, asfrom=True)
+        sql += "USING %s " % compiler.process(self._source_expr, asfrom=True)
+        sql += "ON ( %s ) " % compiler.process(self._on)
+        if self._is_merge_update():
+            sql += self._compile_update(compiler)
+        if self._is_merge_delete():
+            sql += self._compile_delete(compiler)
+        if self._is_merge_insert():
+            sql += self._compile_insert(compiler)
+        return sql
```

Comment:
I guess, you simply extracted the parts into methods, this way better understandable